src/query/sql/src/planner/optimizer/property/constraint.rs

```diff
@@ -146,22 +151,25 @@ impl<'ctx> ScalarVisitor for VariableCollector<'ctx> {
 
 /// Transform a logical expression into a z3 ast.
 /// Will return a Nullable Boolean ast.
-fn transform_logical_expr<'ctx>(ctx: &'ctx Context, scalar: &ScalarExpr) -> Option<Dynamic<'ctx>> {
+fn transform_logical_expr<'ctx>(
+    ctx: &'ctx Context,
+    scalar: &mut ScalarExpr,
```

Comment:
Maybe there is no need to mutate the original `ScalarExpr`. Just returning a rewritten result is enough.

Reply:
In `is_null_reject`, we will call `as_z3_ast` and then call `transform_logical_expr`, if we return rewritten result, we need to create an unnecessary Vec.